R/expr-array.R

```diff
@@ -0,0 +1,418 @@
+  self$`_rexpr`$arr_arg_min() |>
+    wrap()
+}
+
+#' Retrieve the index of the maximum value in every sub-array
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(1:2, 2:1)
+#' )$cast(pl$Array(pl$Int32, 2))
+#' df$with_columns(
+#'   arg_max = pl$col("values")$arr$arg_max()
+#' )
+expr_arr_arg_max <- function() {
+  self$`_rexpr`$arr_arg_max() |>
+    wrap()
+}
+
+#' Evaluate whether all boolean values are true for every sub-array
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(c(TRUE, TRUE), c(FALSE, TRUE), c(FALSE, FALSE), c(NA, NA)),
+#' )$cast(pl$Array(pl$Boolean, 2))
+#' df$with_columns(all = pl$col("values")$arr$all())
+expr_arr_all <- function() {
+  self$`_rexpr`$arr_all() |>
+    wrap()
+}
+
+#' Evaluate whether any boolean value is true for every sub-array
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(c(TRUE, TRUE), c(FALSE, TRUE), c(FALSE, FALSE), c(NA, NA)),
+#' )$cast(pl$Array(pl$Boolean, 2))
+#' df$with_columns(any = pl$col("values")$arr$any())
+expr_arr_any <- function() {
+  self$`_rexpr`$arr_any() |>
+    wrap()
+}
+
+#' Shift values in every sub-array by the given number of indices
+#'
+#' @inheritParams DataFrame_shift
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   values = list(1:3, c(2L, NA, 5L)),
+#'   idx = 1:2,
+#' )$cast(values = pl$Array(pl$Int32, 3))
+#' df$with_columns(
+#'   shift_by_expr = pl$col("values")$arr$shift(pl$col("idx")),
+#'   shift_by_lit = pl$col("values")$arr$shift(2)
+#' )
+expr_arr_shift <- function(n = 1) {
+  self$`_rexpr`$arr_shift(as_polars_expr(n, as_lit = TRUE)$`_rexpr`) |>
+    wrap()
+}
+
+
+#' Convert an Array column into a List column with the same inner data type
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' df <- pl$DataFrame(
+#'   a = list(c(1, 2), c(3, 4))
+#' )$cast(pl$Array(pl$Int8, 2))
+#'
+#' df$with_columns(
+#'   list = pl$col("a")$arr$to_list()
+#' )
+expr_arr_to_list <- function() {
+  self$`_rexpr`$arr_to_list() |>
+    wrap()
+}
+
+
+# TODO-REWRITE: implement this
+# #' Convert array to struct
+# #'
+# #' @inheritParams expr_list_to_struct
+# #'
+# #' @inherit as_polars_expr return
+# #' @examples
+# #' df <- pl$DataFrame(
+# #'   values = list(1:3, c(2L, NA, 5L))
+# #' )$cast(pl$Array(pl$Int32, 3))
+# #' df$with_columns(
+# #'   struct = pl$col("values")$arr$to_struct()
+# #' )
+# #'
+# #' # pass a custom function that will name all fields by adding a prefix
+# #' df2 <- df$with_columns(
+# #'   pl$col("values")$arr$to_struct(
+# #'     fields = \(idx) paste0("col_", idx)
+# #'   )
+# #' )
+# #' df2
+# #'
+# #' df2$unnest()
+# expr_arr_to_struct <- function(fields = NULL) {
+#   self$`_rexpr`$arr_to_struct(fields) |>
+#     wrap()
+# }
+
+#' Count how often a value occurs in every sub-array
+#'
+#' @param element An expression that produces a single value.
```

Comment:
> I think I'm just missing your point, do you have an example of an argument whose type is unclear in this subnamespace? I agree that in other subns it can be hard to tell but here it's mostly straightforward using param description and default values.

I don't think this element actually needs to be a polars expression.

Reply:
Yes it is: https://docs.pola.rs/docs/python/dev/reference/expressions/api/polars.Expr.arr.count_matches.html#polars-expr-arr-count-matches

Reply:
Ok it could be "Can be Expr or sth coercible to an Expr"